lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -644,62 +644,13 @@ checkSpecializationRequirements(ArrayRef<Requirement> Requirements) {
 
 ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
                                      ArrayRef<Requirement> Requirements) {
-  if (shouldNotSpecializeCallee(OrigF))
-    return;
+  assert(!shouldNotSpecializeCallee(OrigF));
 
   // Perform some sanity checks for the requirements
   checkSpecializationRequirements(Requirements);
 
   OriginalF = OrigF;
   ConvertIndirectToDirect = true;
-
-  SILModule &M = OrigF->getModule();
-  auto &Ctx = M.getASTContext();
-
-  auto OrigGenericSig = OrigF->getLoweredFunctionType()->getGenericSignature();
-  auto OrigGenericEnv = OrigF->getGenericEnvironment();
-
-  std::tie(SpecializedGenericEnv, SpecializedGenericSig) =
-      getSignatureWithRequirements(OrigGenericSig, OrigGenericEnv,
-                                   Requirements, M);
-
-  {
-    SmallVector<Substitution, 4> List;
-
-    OrigGenericSig->getSubstitutions(
-      [&](SubstitutableType *type) -> Type {
-        return SpecializedGenericEnv->mapTypeIntoContext(type);
-      },
-      LookUpConformanceInSignature(*SpecializedGenericSig),
-      List);
-    ClonerParamSubs = Ctx.AllocateCopy(List);
-  }
-
-  {
-    SmallVector<Substitution, 4> List;
-
-    SpecializedGenericSig->getSubstitutions(
-      [&](SubstitutableType *type) -> Type {
-        return OrigGenericEnv->mapTypeIntoContext(type);
-      },
-      LookUpConformanceInSignature(*SpecializedGenericSig),
-      List);
-    CallerParamSubs = Ctx.AllocateCopy(List);
-  }
-
-  {
-    CallerInterfaceSubs = OrigGenericSig->getSubstitutionMap(
-      [&](SubstitutableType *type) -> Type {
-        return SpecializedGenericEnv->mapTypeOutOfContext(
-          SpecializedGenericEnv->mapTypeIntoContext(type));
-      },
-      LookUpConformanceInSignature(*SpecializedGenericSig));
-  }
-
-  OriginalParamSubs = CallerParamSubs;
-
-  HasUnboundGenericParams = !SpecializedGenericSig->areAllParamsConcrete();
-  createSubstitutedAndSpecializedTypes();
```

Comment:
Removing all this code from the constructor will break explicit specialization which is used by @_specialize (this constructor is invoked by the EagerSpecializer).